---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
   eval=F,
  echo=TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
library('data.table')
library('knitr')
library('kableExtra')
library('ggplot2')
library('tidyverse')
library('glue')
library("patchwork") 
library("ggthemes")
library("devtools")
library("ggpubr")
library("paletteer")
library("grid")
library("gridExtra")
library("plotly")


require(DT)
require(formatR)
require(waiter)
require(gridExtra)
require(readxl)
require(scales)



require(ggforce)
require(ontologyIndex)
require(ontologySimilarity)
```

# SNPannot

<!-- badges: start -->
<!-- badges: end -->

SNPannot is a package designed to perform an annotation of single nucleotide polymorphisms. 
<br />
<br />
This package has been developed in the context of the CIBERESP Strategic Subprogram Genrisk (https://cancer.genrisk.org/en/).

## Installation

You can install the development version of SNPannot from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("biodama/SNPannot")
```
## Data Source
This set of functions in package accesses data from:

* **NCBI’s dbSNP SNP database**
  + See [dbSNP](https://www.ncbi.nlm.nih.gov/snp/) for more details.
  + Function: **dbSNP_info()**
* **NCBI’s dbSNP Gene database**
  + See [Gene](https://www.ncbi.nlm.nih.gov/gene/) for more details.
  + Function: **snp_gene()**
* **Ontology Resource**
  + See [GO](http://geneontology.org/) for more details.
  + Function: **snp_gene()**

  
## Example 1
### **dbSNP_info()**
Searches the dbSNP database belonging to NCBI in order to annotate the genes belonging to that SNP. According to the SNP name or position, the function annotates the GRCh37, GRCh38 and also SNPs in linkage disequilibrium (LD) [1000Genomes:phase3](https://www.internationalgenome.org/data).

```{r, eval=T,tidy=TRUE, tidy.opts=list(arrow=TRUE, indent=2)}
library(SNPannot)

temp <- data.table::fread("https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000146/ScoringFiles/Harmonized/PGS000146_hmPOS_GRCh37.txt.gz",skip=19)
temp$pos<-paste0(temp$"chr_name",":",temp$"chr_position")
temp<-data.frame(temp)

```

```{r, eval=T,echo=F, tidy.opts=list(arrow=TRUE, indent=2),results='asis'}
temp2<-dplyr::select(temp, rsID,effect_allele,
                     other_allele,allelefrequency_effect,hm_source,hm_rsID,pos)

colnames(temp2)<-c("rsID","effect-allele","other-allele","MAF",
                   "hm-source","hm-rsID","Position")
kable(temp2,align = "c", escape = F, booktabs = T, 
      linesep = "",position="float_rigth")

```

```{r, eval=T,echo=T,warning=FALSE,message=FALSE,tidy=TRUE, tidy.opts=list(arrow=TRUE, indent=2)}
# Searching for information on some SNPs
res_loop<-dbSNP_info(dat=temp[5:10,1],type="rs",p=T,build=37,r2=0.99,pop="EUR")

```

```{r, eval=T,echo=F, tidy.opts=list(arrow=TRUE, indent=2),results='asis'}

colnames(res_loop)<-c("term","rsID","Gene","Alleles","GRCh37","GRCh38","rsLD")
kable(res_loop,align = "c", escape = F, booktabs = T,
      linesep = "",position="float_rigth",responsive=T)
```



## Example 2
### **snp_gene()**

Searches NCBI Gene database and Gene Ontology Resource in order to obtain relevant and specific information about the entered genes. The function can be based on data generated by the *dbSNP_info()*, or on user data.

```{r, eval=T,echo=T,warning=FALSE,message=FALSE,tidy=TRUE}
# Searching for information on some SNPs
snps_info<-dbSNP_info(dat=temp[c(3,10,12:13),1],
                      type="rs",p=T,build=37,r2=0.999,pop="EUR")
genes_info1<-snp_gene(dat=snps_info,type="info",p=T)

```

```{r, eval=T,echo=F, tidy.opts=list(arrow=TRUE, indent=2),results='asis'}

genes_info_1<-dplyr::select(genes_info1,"term","hgnc","entrez_id","gene_des","gene_type","refseq","also")
genes_info_2<-dplyr::select(genes_info1,info,GO_id,GO_des)
colnames(genes_info_1)<-c("term","hgnc","Entrez-id","gene-des","gene-type",
                         "refseq","also")
colnames(genes_info_2)<-c("info","GO-id","GO-des")

kable(genes_info_1,align = "c", escape = F, booktabs = T,
      linesep = "",position="float_rigth")

term<-genes_info_1[,1]
genes_info_2<-cbind(term,genes_info_2)
kable(genes_info_2,align = "c", escape = F, booktabs = T,
      linesep = "",position="float_rigth")

```
## License

[CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/?ref=chooser-v1)

